src/workspace/mod.rs

```diff
@@ -121,10 +123,32 @@ impl Workspace {
         Ok(self)
     }
 
-    /// Generates a package to invoke for generating contract metadata
-    pub(super) fn with_metadata_gen_package(&mut self) -> Result<&mut Self> {
+    /// Generates a package to invoke for generating contract metadata.
+    ///
+    /// The contract metadata will be generated for the supplied `package_name`
+    /// found at `package_path`.
+    pub(super) fn with_metadata_gen_package(
+        &mut self,
+        package_name: String,
+        package_path: PathBuf,
+    ) -> Result<&mut Self> {
+        // We strip the workspace root path from the path where the package is found.
+        // This way we have the relative path of the package in the workspace.
+        let relative_package_path = package_path
+            .strip_prefix(self.workspace_root.clone())
+            .expect("package path must be prefixed with workspace path");
+        // We prepend this path since the metadata generation package will be under
+        // `.ink/metadata_gen/` and we need to traverse up from there.
+        let mut path = PathBuf::from("../../");
```

Comment:
This seems brittle (dependent on the directory structure) , it would be good to encapsulate this somehow in `manifest`.

Reply:
So the reason why I've put this here is that only in this module we have information about the `workspace_root`. In `manifest` we only have information about the local view of a package.

But in order to have our metadata generation crate reference the sub-contract we need to hand the information about the relative path to the sub-contract to the `manifest`. The only other way I see is to add something like `relative_root_path` to `manifest`. Wdyt?

Reply:
> We can't use the absolute path, since we copy the contract to a temp directory.

I think we can make it an absolute path, we have access to the tmp dir path here where we call into `manifest.write` https://github.com/paritytech/cargo-contract/blob/f342f2364f1e7ba8859ca2f1f64bdcd25e203021/src/workspace/mod.rs#L177